<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="A convenience wrapper for differential analysis with LIMMA.
Performs moderated t-tests, F-tests, or linear regression."><title>LIMMA Convenience Wrapper — limma_full • MotrpacRatTraining6moWAT</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="LIMMA Convenience Wrapper — limma_full"><meta property="og:description" content="A convenience wrapper for differential analysis with LIMMA.
Performs moderated t-tests, F-tests, or linear regression."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">MotrpacRatTraining6moWAT</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/MoTrPAC/MotrpacRatTraining6moWAT/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>LIMMA Convenience Wrapper</h1>
      <small class="dont-index">Source: <a href="https://github.com/MoTrPAC/MotrpacRatTraining6moWAT/blob/HEAD/R/limma_full.R" class="external-link"><code>R/limma_full.R</code></a></small>
      <div class="d-none name"><code>limma_full.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A convenience wrapper for differential analysis with LIMMA.
Performs moderated t-tests, F-tests, or linear regression.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">limma_full</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">model.str</span>,</span>
<span>  <span class="va">coef.str</span>,</span>
<span>  <span class="va">contrasts</span>,</span>
<span>  trend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  robust <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  var.group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,</span>
<span>  block <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  adjust.method <span class="op">=</span> <span class="st">"BH"</span>,</span>
<span>  adjust.globally <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p>Object of class
<code><a href="https://rdrr.io/pkg/Biobase/man/class.ExpressionSet.html" class="external-link">ExpressionSet</a></code>. The <code>exprs</code>
slot can be either a matrix of log\(_2\) relative abundances or counts.
If the latter, the limma--edgeR pipeline
(<a href="https://doi.org/10.12688/f1000research.9005.3" class="external-link">https://doi.org/10.12688/f1000research.9005.3</a>) will automatically be
used.</p></dd>


<dt>model.str</dt>
<dd><p>character; formulation of the model (e.g. <code>"~ a + b"</code>).
If <code>contrasts</code> are provided, this should be a no-intercept model (it
should include 0 or -1 as terms). See <code><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></code> for more
details.</p></dd>


<dt>coef.str</dt>
<dd><p>character; coefficient of interest. One of
<code>colnames(pData(object))</code>.</p></dd>


<dt>contrasts</dt>
<dd><p>character; optional contrasts of the form <code>"a-b"</code> to
test.</p></dd>


<dt>trend</dt>
<dd><p>logical; should an intensity-dependent trend be allowed for the
prior variance? If <code>FALSE</code>, then the prior variance is constant.
Alternatively, <code>trend</code> can be a row-wise numeric vector, which will be
used as the covariate for the prior variance. See
<code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></code> for more details.</p></dd>


<dt>robust</dt>
<dd><p>logical; should the estimation of <code>df.prior</code> and
<code>var.prior</code> be robustified against outlier sample variances? See
<code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></code> for more details.</p></dd>


<dt>var.group</dt>
<dd><p>character; the column in <code>pData(object)</code> indicating
groups that will have different relative quality weights. The default
(<code>character(0)</code>) will weight each sample equally (all samples given a
weight of 1). Weights affect the logFC and P.Value columns of the results.</p></dd>


<dt>block</dt>
<dd><p><code>NULL</code> or character; name of a column in
<code>pData(object)</code> specifying a blocking variable. Passed to
<code><a href="https://rdrr.io/pkg/limma/man/dupcor.html" class="external-link">duplicateCorrelation</a></code>. Section 9.7 "Multi-level
Experiments" of the LIMMA User's Guide (see
<code><a href="https://rdrr.io/pkg/limma/man/limmaUsersGuide.html" class="external-link">limmaUsersGuide</a></code>) explains when to use
<code>duplicateCorrelation</code>. Currently ignored if <code>exprs(object)</code> is a
matrix of counts.</p></dd>


<dt>plot</dt>
<dd><p>logical; whether to generate diagnostic plots. If <code>TRUE</code>,
generates a barplot of weights applied to each sample and a plot of the
mean-variance trend using <code><a href="https://rdrr.io/pkg/limma/man/plotSA.html" class="external-link">plotSA</a></code>.</p></dd>


<dt>adjust.method</dt>
<dd><p>method for p-value adjustment. Default is <code>"BH"</code>
(Benjamini-Hochberg), which will control the false discovery rate (FDR).
See <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></code> for details.</p></dd>


<dt>adjust.globally</dt>
<dd><p>logical; should p-values from all contrasts be
adjusted together using <code>adjust.method</code>? Set to <code>FALSE</code> if the
contrasts being tested are not closely related. See Section 13.3 "Multiple
Testing Across Contrasts" of the LIMMA User's Guide
(<code><a href="https://rdrr.io/pkg/limma/man/limmaUsersGuide.html" class="external-link">limmaUsersGuide</a></code>) for more information.</p></dd>


<dt>...</dt>
<dd><p>additional arguments passed to <code><a href="https://rdrr.io/pkg/limma/man/plotSA.html" class="external-link">plotSA</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code>data.frame</code>. Output of <code><a href="https://rdrr.io/pkg/limma/man/toptable.html" class="external-link">topTable</a></code> with
additional columns <code>feature</code>, <code>contrast</code>, and column(s) for the
standard error (SE) of the logFC. All columns from <code>fData</code> are also
included.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>An MDS plot (<code><a href="https://rdrr.io/pkg/limma/man/plotMDS.html" class="external-link">plotMDS</a></code>) is used to determine the
appropriate value of <code>var.group</code>. If samples within phenotype groups
tend to cluster well and have similar dispersions, the default
<code>var.group = character(0)</code> is recommended. If samples within phenotype
groups tend to cluster well, but different groups are more or less
dispersed, it may be beneficial to set <code>var.group</code> to the name of the
phenotype column. If one or more samples tend to cluster poorly with
samples of the same phenotype, it may be beneficial to weight by sample.
That is, set <code>var.group</code> to the name of the column in
<code>pData(object)</code> that uniquely identifies each sample. If variation
between samples or groups of samples is biological rather than technical,
weighting is not recommended.</p>
<p>The plot of the mean-variance trend helps determine whether to fit a trend
line to the prior variance (default is <code>trend = TRUE</code>). It also helps
determine if the prior variance should be robustified against outlier
sample variances (<code>robust = TRUE</code>). See <code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></code>
for more details.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Smyth G. K. (2004). Linear models and empirical bayes methods for
assessing differential expression in microarray experiments.
<em>Statistical applications in genetics and molecular biology, 3</em>,
Article3. <a href="https://doi.org/10.2202/1544-6115.1027" class="external-link">https://doi.org/10.2202/1544-6115.1027</a></p>
<p>Smyth, G. K., Michaud, J., and Scott, H. (2005). The use of within-array
replicate spots for assessing differential expression in microarray
experiments. <em>Bioinformatics 21</em>(9), 2067--2075.
<a href="http://bioinformatics.oxfordjournals.org/content/21/9/2067" class="external-link">http://bioinformatics.oxfordjournals.org/content/21/9/2067</a> Preprint
with corrections: <a href="http://www.statsci.org/smyth/pubs/dupcor.pdf" class="external-link">http://www.statsci.org/smyth/pubs/dupcor.pdf</a></p>
<p>Liu, R., Holik, A. Z., Su, S., Jansz, N., Chen, K., Leong, H. S., Blewitt,
M. E., Asselin-Labat, M. L., Smyth, G. K., &amp; Ritchie, M. E. (2015). Why
weight? Modelling sample and observational level variability improves power
in RNA-seq analyses. <em>Nucleic acids research, 43</em>(15), e97.
<a href="https://doi.org/10.1093/nar/gkv412" class="external-link">https://doi.org/10.1093/nar/gkv412</a></p>
<p>Law, C. W., Alhamdoosh, M., Su, S., Dong, X., Tian, L., Smyth, G. K., &amp;
Ritchie, M. E. (2016). RNA-seq analysis is easy as 1-2-3 with limma, Glimma
and edgeR. <em>F1000Research, 5</em>, ISCB Comm J-1408.
<a href="https://doi.org/10.12688/f1000research.9005.3" class="external-link">https://doi.org/10.12688/f1000research.9005.3</a></p>
<p>Phipson, B., Lee, S., Majewski, I. J., Alexander, W. S., &amp; Smyth, G. K.
(2016). ROBUST HYPERPARAMETER ESTIMATION PROTECTS AGAINST HYPERVARIABLE
GENES AND IMPROVES POWER TO DETECT DIFFERENTIAL EXPRESSION. <em>The
annals of applied statistics, 10</em>(2), 946--963.
<a href="https://doi.org/10.1214/16-AOAS920" class="external-link">https://doi.org/10.1214/16-AOAS920</a></p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf" class="external-link">LIMMA
User's Guide</a></p>
<p><a href="https://support.bioconductor.org/p/125489/#125602" class="external-link">When to use
<code>duplicateCorrelation</code></a></p>
<p><a href="https://online.stat.psu.edu/stat555/node/1/" class="external-link">PennState STAT 555 -
Statistical Analysis of Genomics Data</a></p>
<p><a href="http://varianceexplained.org/statistics/interpreting-pvalue-histogram/" class="external-link">How
to interpret a p-value histogram</a></p>
<p>Breheny, P., Stromberg, A., &amp; Lambert, J. (2018). <em>p</em>-Value Histograms:
Inference and Diagnostics. <em>High-throughput, 7</em>(3), 23.
<a href="https://doi.org/10.3390/ht7030023" class="external-link">https://doi.org/10.3390/ht7030023</a></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Tyler Sagendorf, Vlad Petyuk</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Tyler Sagendorf.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

